### PR TITLE
Crud

### PR DESCRIPTION
**Pull request by TiannanXu**

### CODE REVIEW

File: GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+    {
+        private readonly IWebHostEnvironment _webHostEnvironment;
+
+        public ItemRepositoryController(AppDbContext context,IWebHostEnvironment webHostEnvironment) : base(context) 
+        {
+            _webHostEnvironment = webHostEnvironment;
+        }
+
+
+        public async Task<IActionResult> ItemRepositoryCollection([Bind("PageNumber,SelectedOrder")] CollectionViewModel model)
```
Comment (ExploreNcrack) [inline GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs]: should call it just Index

File: GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+                .AsNoTracking();
+
+            var items = allItems.OrderItemsBy(model.SelectedOrder);
+            await model.PaginateItems(items);
+
+            return View(model);
+        }
+
+        [HttpGet]
+        public IActionResult ItemCreate(ItemViewModel model)
```
Comment (ExploreNcrack) [inline GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs]: Just call it Create()

File: GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+using static GunplaOnlineShop.QueryObjects.ItemSort;
+using System.IO;
+using Microsoft.AspNetCore.Http;
+using Microsoft.EntityFrameworkCore.Metadata.Internal;
+
+namespace GunplaOnlineShop.Controllers
+{
```
Comment (ExploreNcrack) [inline GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs:24]: Need to decorate with `[Authorize]` not everyone can access these methods, only the logged in user with admin role can

File: GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs (hunk 2 above)
Comment (ExploreNcrack) [inline GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs]: this action only need to return a static page, no user input is needed, thus no model/user-data should take in

File: GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+            return View(model);
+        }
+
+        [HttpPost]
+        public IActionResult Create(ItemViewModel model)
+        {
+            if (ModelState.IsValid)
+            {
+                Item newItem = new Item();
+                if (_context.Items.Any(c => c.Name == model.Name)) return View("ItemCreate", model);
```
Comment (ExploreNcrack) [inline GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs]: if item with this name exists why return the same create view???

File: GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+                        String uploadsFolder = Path.Combine(_webHostEnvironment.WebRootPath, "images/cover");
+                        var uniqueFileName = Guid.NewGuid().ToString() + "_" + model.CoverPhoto.FileName;
+                        var filePath = Path.Combine(uploadsFolder, uniqueFileName);
+                        using (var fs = new FileStream(filePath, FileMode.Create))
+                        {
+                            model.CoverPhoto.CopyTo(fs);
+                        }
+                        photos.Add(new Photo() { Name = uniqueFileName, Url = "~/images/cover/" + uniqueFileName, FilePath = filePath });
+                    }
+                    if (model.GalleryPhotos != null && model.GalleryPhotos.Count > 0)
```
Comment (ExploreNcrack) [inline GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs:104]: Gallery photo upload is not working properly

File: GunplaOnlineShop/GunplaOnlineShop/Views/ItemRepository/ItemCreate.cshtml [hunk 6]
```diff
@@ -0,0 +1,86 @@
+﻿@model GunplaOnlineShop.ViewModels.ItemViewModel
```
Comment (ExploreNcrack) [inline GunplaOnlineShop/GunplaOnlineShop/Views/ItemRepository/ItemCreate.cshtml]: need to have a back link

File: GunplaOnlineShop/GunplaOnlineShop/Views/ItemRepository/ItemEdit.cshtml [hunk 7]
```diff
@@ -0,0 +1,55 @@
+﻿@model GunplaOnlineShop.ViewModels.ItemViewModel
```
Comment (ExploreNcrack) [inline GunplaOnlineShop/GunplaOnlineShop/Views/ItemRepository/ItemEdit.cshtml]: need to have a back link

File: GunplaOnlineShop/GunplaOnlineShop/Views/ItemRepository/ItemEdit.cshtml [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+                            <input type="hidden" asp-for="CategoryList[i].cateName" />
+                        </li>
+                    }
+                </ul>
+                <div class="form-group">
+                    <label asp-for="Description"></label>
+                    <textarea asp-for="Description" class="form-control" rows="3"></textarea>
+                    <span asp-validation-for="Description" class="text-danger"></span>
+                </div>
+                <button type="submit" class="btn btn-primary btn-lg">Edit</button>
```
Comment (ExploreNcrack) [inline GunplaOnlineShop/GunplaOnlineShop/Views/ItemRepository/ItemEdit.cshtml]: Should call it `Save`

**Review by ExploreNcrack — CHANGES_REQUESTED** [review]

File: GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+                categoryList.Add(new CategoryCheck() { cateName = cate.Name, cateId = cate.Id });
+            }
+
+            model.CategoryList = categoryList;
+
+            return View(model);
+        }
+
+        [HttpPost]
+        public IActionResult Create(ItemViewModel model)
```
Comment (ExploreNcrack) [inline GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs:77]: `[ValidateAntiForgeryToken]`

File: GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+                _context.Items.Add(newItem);
+                _context.SaveChanges();
+                return RedirectToAction("ItemRepositoryCollection", "ItemRepository");
+
+            }
+            return View("ItemCreate", model);
+        }
+
+        [HttpPost]
+        public IActionResult Delete(int? id)
```
Comment (ExploreNcrack) [inline GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs:153]: `[ValidateAntiForgeryToken]`

File: GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+                    _context.Remove(photo);
+                }
+            }
+            _context.SaveChanges();
+
+            return RedirectToAction("ItemRepositoryCollection", "ItemRepository");
+        }
+
+        [HttpGet]
+        public IActionResult ItemEdit(int id)
```
Comment (ExploreNcrack) [inline GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs]: `[ValidateAntiForgeryToken]`

**Review by ExploreNcrack — CHANGES_REQUESTED** [review]

File: GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs (hunk 5 above)
Comment (TiannanXu) [reply GunplaOnlineShop/GunplaOnlineShop/Controllers/ItemRepositoryController.cs:104]: Not sure the problem, it works just fine on my PC

**Pull request merged**